### PR TITLE
Manage user data

### PR DESCRIPTION
**Pull request by zaneb**
https://issues.redhat.com/browse/METAL-74

Related issue: #683
Enhancement: https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/user-data-secret-managed.md#machines-user-data-managed-by-the-mco

This change was previously committed in #1792, but reverted again in #2126. With the installer change openshift/installer#4413 in place, [Bug 1881703](https://bugzilla.redhat.com/show_bug.cgi?id=1881703) is now resolved and it is safe to re-add this code.

This is now required for the enhancement https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/coreos-image-in-release.md#implementation-detailsnotesconstraints

### CODE REVIEW

File: pkg/controller/common/helpers.go
Comment (cgwalters) [inline pkg/controller/common/helpers.go:139]: Surely we should have this start out as spec3 now?

**Review by cgwalters — APPROVED** [review]
OK I re-skimmed the previous discussion.  I have to admit I can just like *barely* keep all this in my head and it will surely get context switched out again.  But your commit message was great in linking to all of the relevant bits.

So...as far as I understand things, the next required step here would be for the installer to switch over to generating machinesets to point to `-user-data-managed` if it exists?

Marking approved from my PoV but this also requires an MCO team approval I'd say.

File: pkg/controller/common/helpers.go
Comment (zaneb) [reply pkg/controller/common/helpers.go:139]: A yep, I did a cherry-pick of the original patch instead of a revert of the revert, so I missed 22b8b24a39bfdcfab82e47ba9ebb3dc22e0e716d (and f6b693805f4fa941a74096ccad7e393878014105) on the first pass. Will add those in.

**Comment by zaneb** [comment]
> So...as far as I understand things, the next required step here would be for the installer to switch over to generating machinesets to point to `-user-data-managed` if it exists?

Yeah, I will need to follow up by also unreverting https://github.com/openshift/installer/pull/4228 and https://github.com/openshift/hive/pull/1166.

**Comment by cgwalters** [comment]
> Yeah, I will need to follow up by also unreverting openshift/installer#4228 and openshift/hive#1166.

Oh man, I hadn't realized hive needed changes for this too.  It'd be really nice to somehow have that share code with openshift/installer.

And just for completeness, now that https://github.com/openshift/hypershift exists does it need updating too?  A quick [code search](https://github.com/openshift/hypershift/search?q=user-data&type=commits] turns up https://github.com/openshift/hypershift/commit/d98c7a7ae251a62c682a573a6ad43261d2da5a81 which seems to have dropped that dependency?

**Comment by zaneb** [comment]
> Oh man, I hadn't realized hive needed changes for this too. It'd be really nice to somehow have that share code with openshift/installer.

Me neither, I just found it looking at that installer PR.
I don't think it can share code because it needs to do different things based on the version of the cluster being installed :(

> And just for completeness, now that https://github.com/openshift/hypershift exists does it need updating too?

As far as I can tell, no. I don't see any reference to [master|worker]-user-data in that repo.

**Comment by zaneb** [comment]
OK, I think it was actually a mistake for the installer change to modify the MachineSets for every platform - to date all of those MachineSets have fixed images that never change, so their user data should never change. The only platform we should switch over the _worker_ image names for is baremetal.

The baremetal platform doesn't use MachinePools in Hive, so if we limit the change to that then there is nothing to do in Hive.

**Comment by hardys** [comment]
/lgtm

So just to clarify - the reason this got reverted is it broke the installer interface that allows users to customize the pointer ignition.  

As mentioned in the description I subsequently made changes to the installer which ensure customizations to the pointer config are translated to a MachineConfig, so it should be safe to restore this AFAICS.

/cc @yuqi-zhang @runcom

**Comment by zaneb** [comment]
It appears to me that the remaining tests here are perma-failing for reasons unrelated to this patch.

/assign @sinnykumari
/retest

File: pkg/controller/common/controller_context.go
Comment (yuqi-zhang) [inline pkg/controller/common/controller_context.go:79]: If the user wants to customize their pointer config, they should not customize the `managed` secret, but instead create a new one to use, right?

**Review by yuqi-zhang — COMMENTED** [review]
Sorry I missed this originally. The code looks fine on a first pass (most of it is ported from the original PR). At a high level I have a few questions:

1. If the user wants to customize their pointer config, they should not customize the `managed` secret, but instead create a new one to use, right?
2. Does this assumption still hold true: for existing install upgrading, we will simply have parallel worker/master/worker-managed/master-managed secrets. For new installs, we will only have `managed` secrets?
3. We have been telling users to do more custom ignition for e.g. partitioning different types of disks for different nodes in the same pool. If we go forward with this, we should be careful to note this change via relevant openshift doc sections
4. Reading the enhancement and issue, there is no set release version on them. Is this required for 4.10?

**Comment by zaneb** [comment]
> Reading the enhancement and issue, there is no set release version on them. Is this required for 4.10?

In 4.10, on the baremetal platform we will begin creating MachineSets that always use the latest CoreOS (rather than the one from install time) for provisioning new Machines.

AIUI this means we should also switch them to the managed userdata, so that if in future the ignition format changes in a non-upward-compatible way again it will still use a version that matches the CoreOS version. If we don't land it in 4.10 then we'd have to rewrite the MachineSets in some later release if this did happen. @cgwalters is that right?

> Does this assumption still hold true: for existing install upgrading, we will simply have parallel worker/master/worker-managed/master-managed secrets.

Yes.

> For new installs, we will only have managed secrets?

As far as workers go, I think we should do this only on the baremetal platform, because all other platforms continue to use the CoreOS image from cluster install time in their MachineSets, so they should also use the user data from cluster install time.
For masters, it's less clear-cut what the right thing is, since there are no MachineSets.

**Comment by yuqi-zhang** [comment]
> As far as workers go, I think we should do this only on the baremetal platform, because all other platforms continue to use the CoreOS image from cluster install time in their MachineSets, so they should also use the user data from cluster install time.
For masters, it's less clear-cut what the right thing is, since there are no MachineSets.

So basically we won't switch over openshift/installer to using the `-managed` secrets just yet, and only have this as an availability for baremetal workers.

Do you plan on making that (baremetal worker using -managed) the default via an installer PR?

**Comment by zaneb** [comment]
> Do you plan on making that (baremetal worker using -managed) the default via an installer PR?

[Yes](https://github.com/zaneb/openshift-installer/commits/mco-user-data-managed).

**Review by yuqi-zhang — APPROVED** [review]
Brought up a cluster to check, can confirm the cluster has 4 secrets (pool + pool-managed) by default with the same contents. PR should be fine to merge, but we should probably consider the best way to introduce this to newly installed clusters.

Also waiting to see if others have more feedback

**Comment by zaneb** [comment]
Is it worth overriding the e2e-agnostic-upgrade job, which seems to be broken eveywhere for everyone?

**Comment by zaneb** [comment]
Installer changes for this are in https://github.com/openshift/installer/pull/5473 (specifically, the [last two commits](https://github.com/openshift/installer/pull/5473/files/781feee735b5d89d2cfc627d449a004878970103..6fbad0dc72e680aa7eec2e8b86f2fac1402ee935)), and affect only the baremetal platform.

**Pull request merged**
